src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/KestrelHttpParser.cs

```diff
@@ -473,13 +384,244 @@ public KestrelHttpParser(IKestrelTrace log)
 
                 var nameBuffer = span.Slice(nameStart, nameEnd - nameStart);
                 var valueBuffer = span.Slice(valueStart, valueEnd - valueStart);
-                consumedBytes += headerLineLength;
+                consumedBytes += span.Length;
 
                 handler.OnHeader(nameBuffer, valueBuffer);
+
                 consumed = reader.Cursor;
             }
         }
 
+        private unsafe bool TakeMessageHeadersSingleSpan<T>(T handler, Span<byte> headersSpan, out int consumedBytes) where T : IHttpHeadersHandler
+        {
+            consumedBytes = 0;
+
+            var remaining = headersSpan.Length;
+            var index = 0;
+            fixed (byte* data = &headersSpan.DangerousGetPinnableReference())
+            {
+                while (true)
+                {
+                    if (remaining < 2)
+                    {
+                        return false;
+                    }
+
+                    var ch1 = data[index];
```

Comment:
Oh, now I get it. Makes me sad. I feel this over complicates the main parser making it hard to work on it.